New blog article about high load in Java applications (#48)
qaware/qaware-blog-source: content/posts/java-service-queues.md

---
title: "Don't Crack Under Pressure: Java Microservices and the Battle for Stability Under High Load"
date: 2023-06-12T20:52:33+02:00
author: "[Sebastian Macke](https://github.com/s-macke)"
type: "post"
tags: ["Java", "Web Framework", "REST", "Services", "Thread", "Queue"]
image: "java-service-queues/connections.png"
draft: true
summary: Explore the challenges of Java microservices under high load conditions.
---

In recent years, I have been working for a large, well-established corporation with an extensive 
IT ecosystem that spans several decades, including various infrastructure, software, and cloud environments, 
with Java being the prominent language. Maintaining seamless operations and ensuring 24/7 availability 
of all these systems is an immense challenge.

As site reliability engineer, I have frequently dealt with outages. Although there 

> could be countless reasons for outages, one technical issue that I find particularly bothersome is *high load* in Java servcices 

Reviewer: ```suggestion
could be countless reasons for outages, one technical issue that I find particularly bothersome is *high load* in Java services 
```

> and how to fall over safely. Especially when there is no ressource shortage at all.

Reviewer: ```suggestion
and how to fail safely. Especially when there is no resource shortage at all.
```

This article is aimed at those who want to improve the reliability of their Java-based web application services 
and those interested in gaining insights into load testing their services.

# A simple Java Microservice

> Let's try an experiment together. Take a Web Application Java Framework. Any framework will do. [Spring Boot](https://spring.io/projects/spring-boot) ? Ok.

Reviewer: ```suggestion
Let's try an experiment together. Take a Web Application Java Framework. Any framework will do. [Spring Boot](https://spring.io/projects/spring-boot)? Ok.
```

This test runs on the same machine as the servcice and sends 5000 requests per second.

{{< animated static="/images/java-service-queues/sb_0s_5000req_play.svg"              anim="/images/java-service-queues/sb_0s_5000req.svg" >}}

The success rate is 100%, and the result always has a status code of 200. Even with such a high number of requests, the average response time for each request is less than 1ms.

# Adding a blocking action

Now, let's discuss a more realistic situation. Typically, a request involves additional actions like authentication, caching, file system access, database queries, and other backend calls. 
These actions delay the current request until the result is ready.

It can take anywhere from a few milliseconds to several seconds, depending on the complexity and specific scenarios. In some cases, a request 
may take up to dozens of seconds to get a response. While this is not ideal for the user, it is generally acceptable. 
After all, we don't have the resources of giants like Google or Amazon to optimize every single request to its full potential.

To simulate a blocking action, we can simply add a *sleep* command to our code.

> ```log

Reviewer: ```suggestion
```Java
```

First, let's consider worker threads. Should we increase their number? Yes, we can do that. 
However, remember that these threads are managed by the kernel, which is a shared resource 
and responsible for the stability of your system. Overburdening the kernel with thread organization tasks is not recommended. 
Have you ever heard of a [fork bomb](https://en.wikipedia.org/wiki/Fork_bomb)?

Unfortunately, I have not found any information on a reasonable maximum capacity for the kernel. 
Under laboratory conditions, it can handle 10,000 threads without any problem. 
However, I am not sure about its capacity in a shared Kubernetes cluster or in a 24/7 scenario. 
Server operating systems can also impose limits on the maximum number of threads. 
For example, some Linux server distributions set the default limit to a low number, such as 512.

In any case, it is impossible to predict the optimal number of worker threads for every scenario. 
This makes it a parameter that we want to eliminate in the long term (see Reactive and Virtual Threads below).

Now let's consider the queue. Should we limit its size? *Yes, definitely*. Reasonable values can be found in the book:

> {{< img src="/images/java-service-queues/sre_book.jpg" >}}

Reviewer: Hier fehlt im Text der Name des Buches, oder zumindest ein Alt-Text

In no line of the book do they mention unbounded queues.

The worst recommendation I've come across is in the documentation of the [Quarkus](https://quarkus.io/) framework:
{{< img src="/images/java-service-queues/quarkus_queue_suggestion.png" >}}

The discrepancy between the Java Framework developers and Site Reliability Engineers is just 0 vs. infinite. 
They don't even try to give description of the parameter. As if everyone working with microservices is aware of the implications.

Do framework developers even test their services in real-world conditions?

You might think that web frameworks would have reasonable default settings 
in place and provide guidance for configuring services for production use. 
Ideally, every service developer would follow these best practices.
Well, the existence of this article suggests that neither is the case.

# Improving the Spring Boot configuration

<!-- Production readiness or tuning guides. What should we set here? -->

Here's what we want to achieve:
- The service should always be up, running, and responsive.
- If the maximum number of requests is reached, the service should log this issue.

> - The service should stop further requests with an approprite error.

Reviewer: ```suggestion
- The service should stop further requests with an appropriate error.
```

readExecutor@29fa40c9{reserved=5/12,pending=0}] rejected org.eclipse.jetty.io.ManagedSelector$DestroyEndPoint@71038c20
2023-06-11 17:46:49.310  WARN 29714 --- [qtp466319810-37] o.e.jetty.util.thread.QueuedThreadPool   : QueuedThreadPool[qtp466319810]@1bcb79c2{STARTED,8<=200<=200,i=178,r=-1,q=0}[ReservedTh
readExecutor@29fa40c9{reserved=5/12,pending=0}] rejected org.eclipse.jetty.io.ManagedSelector$DestroyEndPoint@242e0902
2023-06-11 17:46:49.310  WARN 29714 --- [qtp466319810-19] o.e.jetty.util.thread.QueuedThreadPool   : QueuedThreadPool[qtp466319810]@1bcb79c2{STARTED,8<=200<=200,i=178,r=-1,q=0}[ReservedTh
readExecutor@29fa40c9{reserved=5/12,pending=0}] rejected org.eclipse.jetty.io.ManagedSelector$DestroyEndPoint@6950a041
2023-06-11 17:46:49.310  WARN 29714 --- [tp466319810-168] o.e.jetty.util.thread.QueuedThreadPool   : QueuedThreadPool[qtp466319810]@1bcb79c2{STARTED,8<=200<=200,i=177,r=-1,q=0}[ReservedTh
readExecutor@29fa40c9{reserved=5/12,pending=0}] rejected org.eclipse.jetty.io.ManagedSelector$DestroyEndPoint@a68223f
```

However, the service only responds with an abrupt disconnect, not an error message for the client. This issue cannot be easily 

> fixed in Java and requires a complete change in the internal workings, as explained in the Reactive chapter.

Reviewer: Welches Chapter? Im O’Reilly-Buch?

The framework [Quarkus](https://quarkus.io/) has addressed this problem and returns actually an error code when the queue is limited.

{{< animated static="/images/java-service-queues/quarkus_1s_250req_q1_play.svg"    anim="/images/java-service-queues/quarkus_1s_250req_q1.svg" >}}

The `500 Internal Server Error` error code can be changed to a more reasonable `503 Service Unavailable` error.

This is probably the best-case scenario for Java services currently. The service remains responsive, and when worker threads are fully occupied, it returns an error code.

But there is more.

# The importance of Circuit Breakers

> Once you understand the behaviour of Java Micro Services under high load you understand the need for circuit breakers in such frameworks.

Reviewer: ```suggestion
Once you understand the behaviour of Java Microservices under high load you understand the need for circuit breakers in such frameworks.
```
